subsys/net/lib/azure_fota/azure_fota.c

```diff
@@ -0,0 +1,277 @@
+
+/* Enum to keep the fota status */
+enum fota_status {
+	STATUS_NONE = 0,
+	STATUS_DOWNLOADING,
+	STATUS_DOWNLOAD_DONE,
+	STATUS_DOWNLOAD_ERROR,
+	STATUS_APPLY_UPDATE,
+	STATUS_REBOOT,
+};
+
+/* Static variables */
+static azure_fota_callback_t callback;
+static size_t download_progress;
+static enum fota_status fota_state;
+static char hostname[CONFIG_AZURE_FOTA_HOSTNAME_MAX_LEN];
+static char file_path[CONFIG_AZURE_FOTA_FILE_PATH_MAX_LEN];
+
+/* Static functions */
+
+static void fota_dl_handler(const struct fota_download_evt *fota_evt)
+{
+	struct azure_fota_event evt;
+
+	switch (fota_evt->id) {
+	case FOTA_DOWNLOAD_EVT_FINISHED:
+		LOG_INF("FOTA download completed evt received");
+
+		fota_state = STATUS_DOWNLOAD_DONE;
+
+		evt.type = AZURE_FOTA_EVT_DL_PROGRESS;
+		evt.dl.progress = AZURE_FOTA_EVT_DL_COMPLETE_VAL;
+		callback(&evt);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERASE_PENDING:
+		evt.type = AZURE_FOTA_EVT_ERASE_PENDING;
+		callback(&evt);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERASE_DONE:
+		evt.type = AZURE_FOTA_EVT_ERASE_DONE;
+		callback(&evt);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERROR:
+		LOG_ERR("FOTA download failed, report back");
+		fota_state = STATUS_NONE;
+		evt.type = AZURE_FOTA_EVT_ERROR;
+		callback(&evt);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_PROGRESS:
+		/* Only if CONFIG_FOTA_DOWNLOAD_PROGRESS_EVT is enabled */
+		download_progress = fota_evt->progress;
+		evt.type = AZURE_FOTA_EVT_DL_PROGRESS;
+		evt.dl.progress = download_progress;
+		callback(&evt);
+		break;
+	default:
+		LOG_WRN("Unrecognized Azure FOTA event type: %d",
+			fota_evt->id);
+		break;
+	}
+}
+
+/* @brief Extracts firmware image information from device twin JSON document.
+ *
+ * @retval 0 if successfully parsed, hostname and file_path populated.
+ * @retval -ENOMSG Could not parse incoming buffer as JSON.
+ * @retval -ENOMEM Could not find object, or insufficient memory to allocate it.
+ */
+static int extract_fw_details(const char *msg)
+{
+	int err = -ENOMEM;
+	struct cJSON *root_obj, *desired_obj, *fw_obj, *fw_version_obj,
+		*fw_location_obj, *fw_check_obj, *fw_hostname_obj, *fw_path_obj;
+	uint32_t fw_version, fw_check;
+
+	/* The properties.desired.firmware object has the excpected structure:
+	 *	properties.desired.firmware : {
+	 *		fwVersion : <version>,
+	 *		fwLocation : {
+	 *			host : <hostname>,
+	 *			path : <file path>
+	 *		},
+	 *		fwCheckValue : <value>
+	 *	}
+	 */
+
+	root_obj = cJSON_Parse(msg);
+	if (root_obj == NULL) {
+		printk("Could not parse message as JSON\n");
+		return -ENOMSG;
+	}
+
+	/* Desired object */
+	desired_obj = cJSON_GetObjectItemCaseSensitive(root_obj, "desired");
+	if (desired_obj == NULL) {
+		/* The received object can be the 'desired' object only, or
+		 * the whole device twin properties object. If there is no
+		 * 'desired' object to be found, assume it's the root object.
+		 */
+		LOG_DBG("No 'desired' object found, assuming it's root object");
+		desired_obj = root_obj;
```

Comment:
Storing data on the Device Twin root is an undocumented behaviour: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-device-twins ... so this case should not be handled.

Reply:
The "root" in this case refers to the root of the received object and not the device twin per se. When the device twin `desired` object is updated, only the `desired` object is received on `$iothub/twin/PATCH/properties/desired/#`, while if the device twin is explicitly requested, the `properties` object of the device twin is received on `$iothub/twin/res/#`. So that's why we need to look for the `firmware` object in the `desired` object it that exists, or in the root of the received object if `desired` doesn't exist.

This should probably be made clearer in the comment in the code, and possibly also have a parameter in `azure_fota_msg_process` to indicate where to look for the `firmware` object.

Reply:
Ah, after double checking the comment I see that it says that.